### DISCUSSION TITLE
Estimation of single-trial activity

### DISCUSSION THREAD

**Discussion by abelowska** [unfoldtoolbox/Unfold.jl#209] [Q&A] [answered]
Hi all,

I am wondering if it is possible to estimate single-trial ERPs after correction, specifically to disentangle stimuli-related and response-related brain activity.

Here is my model specification (I am calling Julia from Python):

```
# Specify model formula
jl.seval("bf_stimuli = firbasis(τ = (-0.3, 0.5), sfreq = 1024)")
jl.seval("bf_response = firbasis(τ = (-0.3, 0.5), sfreq = 1024)")

jl.seval("f = @formula 0 ~ 1")
bfDict = jl.seval("[ \"stimuli\" => (f, bf_stimuli), \"response\" => (f, bf_response) ]")

type_column = jl.seval("Vector{String}")(events_df['type'].tolist())
latency_column = jl.seval("Vector{Int64}")(events_df['latency'].tolist())

# Create the Julia DataFrame
events_df_jl = jl.DataFrame(type=type_column, latency=latency_column)

# Fit Unfold model
m = Unfold.fit(
    Unfold.UnfoldModel,
    bfDict,
    events_df_jl,
    data,
    eventcolumn = "type",
)
```
Where `data` is the continuous EEG signal with a shape of `(1, 935936)`, sampled at 1024 Hz. Using `Unfold.coeftable(m)`, I can extract rERPs for each type of event.

My question is: is there any way to use the estimated coefficients/rERPs to disentangle stimuli-related single-trial ERPs and response-related single-trial ERPs? When calling `predicted_data = Unfold.predict(m, overlap=False)`, I get single-trials, but all have exactly the same values. I suppose this makes sense because I only estimate the intercepts for the effects.

Do you have any recommendations on how to obtain these corrected single-trial ERPs?

Thank you!

**Comment by behinger** [comment]
Hi there,
to confirm I understood what you need: Single trials, without the (modelled) overlap of neighbouring events?

If yes, you are in luck, this is implemented

Cheerio, Bene

**Reply by abelowska** [reply]
Exactly :)

**Reply by behinger** [reply, marked as answer]
Great!

I recommend the internal function "singletrial" from UnfoldDecode.jl
# get overlap free single trails 
`X_train = singletrials(data, uf_model, events, "response", "type" )`

Just in case you want to do decoding on overlap corrected trials, that package would already implement one algorithm to do so. One issue: it is not yet registered, you have to install it via `Pkg.add(url="https://github.com/unfoldtoolbox/UnfoldDecode.jl")`. Unfold decode is unfortunately not yet as polished as unfold, but what you want to do is also not the standard case so that is fine I guess :-)

If you want to put the pieces together yourself, you can use predict with keep/remove basis to exclude/keep certain eventtypes. If you provide to which events to epoch the data at the same time, it will even do it in a memory efficient way. Here is the docs for that (not yet a tutorial, didn't have time yet)
https://github.com/unfoldtoolbox/Unfold.jl/blob/ce2b36f93ecd374ff8cb4192b3de983a7b16c883/src/predict.jl#L160-L191

Ok, long answer, maybe something helpful for you. If you can say more on your use case I might be able to put out a short tutorial for that.
